### PR TITLE
feat(slider): added slider component

### PR DESCRIPTION
**Pull request by mcoker**
fixes https://github.com/patternfly/patternfly/issues/296

**SLIDER LINK - https://patternfly-pr-3711.surge.sh/components/slider**

Designs - https://www.sketch.com/s/6f81d099-0c11-407a-8962-5672df30dcf2/a/j5KAvW

WAI-ARIA example - https://www.w3.org/TR/wai-aria-practices/examples/slider/slider-1.html

### CODE REVIEW

**Review by mcarrano — CHANGES_REQUESTED** [review]
These look good @mcoker . I'll defer to @maryshak1996 and @mceledonia on the visual details, but I did wonder if we can refactor the examples a bit. IMO, the continuous slider is the most common use case. I would either not have an example labeled "Basic" and just have Continuous and Discrete as in the design spec or change the basic example to show a continuous slider with labels at either end as to me this would be the most simple/common. What do you think?

**Review by maryshak1996 — APPROVED** [review]

**Comment by maryshak1996** [comment]
> These look good @mcoker . I'll defer to @maryshak1996 and @mceledonia on the visual details, but I did wonder if we can refactor the examples a bit. IMO, the continuous slider is the most common use case. I would either not have an example labeled "Basic" and just have Continuous and Discrete as in the design spec or change the basic example to show a continuous slider with labels at either end as to me this would be the most simple/common. What do you think?

@mcarrano I don't have a real preference as to which variant should be displayed first, but I think that you bring up a good point about the naming. I think the Continuous and Discrete is more accurate than Continuous and Basic (basic implies that it's the default choice, but I don't know that we want to be making that type of preference suggestion).

File: src/patternfly/components/Slider/slider.scss [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  --pf-c-slider__thumb--focus--BoxShadow: var(--pf-c-slider__thumb--BoxShadow--base);
+  --pf-c-slider__thumb--active--BoxShadow:
+    var(--pf-c-slider__thumb--BoxShadow--base),
+    0 0 2px 5px var(--pf-global--active-color--200);
+
+  // value
+  --pf-c-slider__value--MarginLeft: var(--pf-global--spacer--md);
+  --pf-c-slider__value--c-form-control--width-base: 3.5ch;
+  --pf-c-slider__value--c-form-control--width-chars: 2;
+  --pf-c-slider__value--c-form-control--Width: calc(var(--pf-c-slider__value--c-form-control--width-base) + (var(--pf-c-slider__value--c-form-control--width-chars) * 1ch));
```
Comment (mattnolting) [inline src/patternfly/components/Slider/slider.scss:53]: I think you need to add `+ (var(--pf-global--spacer--sm) * 2)` or the equivalent of `--pf-c-form-control--inset--base`. A 100% value doesn't fit for me

File: src/patternfly/components/Slider/examples/Slider.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+```hbs
+{{#> slider
+  slider--value-min="0"
+  slider--value-max="100"
+  slider--value-now="50"
+  slider--attribute='style="--pf-c-slider--value: 50%;"'
+  }}
+  {{#> slider-actions}}
+    {{#> button button--modifier="pf-m-plain" button--attribute='aria-label="Minus"'}}
+      <i class="fas fa-minus" aria-hidden="true"></i>
```
Comment (mattnolting) [inline src/patternfly/components/Slider/examples/Slider.md:178]:
```suggestion
      <i class="fas fa-minus fa-fw" aria-hidden="true"></i>
```

**Review by christiemolloy — COMMENTED** [review]
When i drag the slider circle, should all the numbers highlight like this?
Screen Shot 2020-12-02 at 11 08 34 AM

File: src/patternfly/components/Slider/examples/Slider.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    {{/button}}
+  {{/slider-actions}}
+  {{#> slider-main}}
+    {{> slider-rail}}
+    {{> slider-thumb}}
+  {{/slider-main}}
+  {{#> slider-actions}}
+    {{#> button button--modifier="pf-m-plain" button--attribute='aria-label="Plus"'}}
+      <i class="fas fa-plus" aria-hidden="true"></i>
```
Comment (mattnolting) [inline src/patternfly/components/Slider/examples/Slider.md:187]:
```suggestion
      <i class="fas fa-plus fa-fw" aria-hidden="true"></i>
```

File: src/patternfly/components/Slider/examples/Slider.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    {{#> slider-value slider-value--modifier="pf-m-floating"}}
+      {{#> input-group}}
+        {{> form-control controlType="input" input="true" form-control--attribute=(concat 'disabled type="number" value="' slider--value-now '" aria-label="Slider value input"')}}
+        {{#> input-group-text input-group-text--modifier="pf-m-plain"}}%{{/input-group-text}}
+      {{/input-group}}
+    {{/slider-value}}
+  {{/slider-main}}
+  {{#> slider-actions}}
+    {{#> button button--modifier="pf-m-plain" button--attribute='aria-label="Locked"'}}
+      <i class="fas fa-lock" aria-hidden="true"></i>
```
Comment (mattnolting) [inline src/patternfly/components/Slider/examples/Slider.md:212]:
```suggestion
      <i class="fas fa-lock fa-fw" aria-hidden="true"></i>
```

**Review by christiemolloy — COMMENTED** [review]
Is the slider thumb meant to visually look like this? It looks different to the design

Screen Shot 2020-12-02 at 11 10 08 AM
Screen Shot 2020-12-02 at 11 10 37 AM

File: src/patternfly/components/Slider/slider.scss [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  height: var(--pf-c-slider__rail-track--Height);
+
+  &::before {
+    position: absolute;
+    top: 0;
+    right: 0;
+    bottom: 0;
+    left: 0;
+    content: "";
+    background:
```
Comment (mattnolting) [inline src/patternfly/components/Slider/slider.scss:87]: Beautiful! 🤩

**Comment by christiemolloy** [comment]
Looks like the little squares should actually be circles (based off the design)

Screen Shot 2020-12-02 at 11 12 00 AM

Screen Shot 2020-12-02 at 11 12 09 AM

**Comment by christiemolloy** [comment]
The design seems to have a small margin-left (in orange) which I don't see in the slider.

Screen Shot 2020-12-02 at 11 14 18 AM
Screen Shot 2020-12-02 at 11 14 22 AM

**Review by mattnolting — COMMENTED** [review]
Excellent work! Left a few comments.

**Comment by mcoker** [comment]
> When i drag the slider circle, should all the numbers highlight like this?
> Screen Shot 2020-12-02 at 11 08 34 AM

@christiemolloy probably since you're basically making a selection dragging a static element like that. Same thing happens with the drawer splitter. Once the react component handles the interactivity, that shouldn't happen.

Screen Shot 2020-12-02 at 10 14 32 AM

**Comment by christiemolloy** [comment]
The color for the thumbnail seems to be too dark in the implementation compared to whats in the design. I cant see the Marvel doc so not sure what it should be

Screen Shot 2020-12-02 at 11 18 29 AM

Screen Shot 2020-12-02 at 11 18 32 AM

**Comment by mcoker** [comment]
> Is the slider thumb meant to visually look like this? It looks different to the design
> 
> Screen Shot 2020-12-02 at 11 10 08 AM
> 
> Screen Shot 2020-12-02 at 11 10 37 AM

@maryshak1996 can you verify? I got the styling from 

Screen Shot 2020-12-02 at 10 19 42 AM

File: src/patternfly/components/Slider/slider.scss (hunk 5 above)
Comment (mcoker) [reply src/patternfly/components/Slider/slider.scss:87]: Thanks! And this should still work when we upgraydd the slider to select a range, we'll just introduce more stops.

**Comment by mcoker** [comment]
> Looks like the little squares should actually be circles (based off the design)
> 
> Screen Shot 2020-12-02 at 11 12 00 AM
> 
> Screen Shot 2020-12-02 at 11 12 09 AM

Nice catch, I didn't notice that. @maryshak1996 can you verify the ticks should be circles instead or do you want to leave them as squares?

**Comment by mcoker** [comment]
> The design seems to have a small margin-left (in orange) which I don't see in the slider.
> 
> Screen Shot 2020-12-02 at 11 14 18 AM
> 
> Screen Shot 2020-12-02 at 11 14 22 AM

I talked with @maryshak1996 about this when I began development, and we agreed the tick would be flush left with the track instead of having a small indent. Mary would you mind verifying this looks as you'd expect?

**Comment by christiemolloy** [comment]
Are you going to show an example with the remove button in the slider: https://www.sketch.com/s/6f81d099-0c11-407a-8962-5672df30dcf2/a/JZ2Qx7

**Comment by christiemolloy** [comment]
In the "Actions" examples, do you think it would be beneficial to show these examples under "codepedent sliders" where the text input sits next to the action button ? https://www.sketch.com/s/6f81d099-0c11-407a-8962-5672df30dcf2/a/JZ2Qx7

**Comment by mcoker** [comment]
> The color for the thumbnail seems to be too dark in the implementation compared to whats in the design. I cant see the Marvel doc so not sure what it should be
> 
> Screen Shot 2020-12-02 at 11 18 29 AM
> 
> Screen Shot 2020-12-02 at 11 18 32 AM

It's blue-500 from the screenshot in https://github.com/patternfly/patternfly/pull/3711#issuecomment-737337117. I'll let @maryshak1996 verify.

**Comment by mcoker** [comment]
> Are you going to show an example with the remove button in the slider: https://www.sketch.com/s/6f81d099-0c11-407a-8962-5672df30dcf2/a/JZ2Qx7

@mcarrano @maryshak1996 what do you think? The actions examples (lock, plus/minus) just show an action you could put into the slider. The lock and plus/minus are not specific parts of this component, only the actions. You could replace those actions with any other sort of action you want, so I only showed a couple of examples with different actions and didn't think it was necessary to show the remove button. Maybe that would go in a future demo with the slider used in the context of a form.

**Comment by maryshak1996** [comment]
@christiemolloy those are good catches (clearly stuff I didn't catch while trying to multitask haha). 
I checked my designs and the thumbnail color is too dark (I made a typo on my spec). The circle and the blue hover/drag outline should both use `blue-400` @mcoker do you want me to update my spec, or are you cool with just updating it for your implementation?
And to the comment about the circles vs rectangles for the tick marks -- I think that, if we could get the main bar itself to have rounded edges then yes, circles is the way to go like this: 
Screen Shot 2020-12-02 at 11 31 58 AM

File: src/patternfly/components/Slider/examples/Slider.md [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+cssPrefix: pf-c-slider
+---
+
+## Examples
+### Discrete
+```hbs
+{{#> slider
+  slider--value-min="0"
+  slider--value-max="8"
+  slider--value-now="5"
```
Comment (christiemolloy) [inline src/patternfly/components/Slider/examples/Slider.md:14]: why "now" instead of a word like "current"

File: src/patternfly/components/Slider/examples/Slider.md [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+  slider--attribute='style="--pf-c-slider--value: 62.5%;"'
```
Comment (christiemolloy) [inline src/patternfly/components/Slider/examples/Slider.md:15]: whats this for

**Comment by mcarrano** [comment]
> @mcarrano @maryshak1996 what do you think? The actions examples (lock, plus/minus) just show an action you could put into the slider. The lock and plus/minus are not specific parts of this component, only the actions. You could replace those actions with any other sort of action you want, so I only showed a couple of examples with different actions and didn't think it was necessary to show the remove button. Maybe that would go in a future demo with the slider used in the context of a form.

Agree that it's not necessary to show this example. As long as you can put any action there the current example is fine.

File: src/patternfly/components/Slider/examples/Slider.md [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+  }}
+  {{#> slider-main}}
+    {{> slider-rail}}
+    {{#> slider-steps}}
+      {{> slider-step slider-step--Left="0%" slider-step--label="0" slider-step--IsActive="true"}}
```
Comment (christiemolloy) [inline src/patternfly/components/Slider/examples/Slider.md:20]: does the left value always have to be in % or could it be numerical

Comment (christiemolloy) [reply src/patternfly/components/Slider/examples/Slider.md:20]: Could the left value and the label equal the same here?

File: src/patternfly/components/Slider/slider.scss [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+  --pf-c-slider__rail-track--before--fill--BackgroundColor--gradient-stop: var(--pf-c-slider--value);
+
+  // steps
+  --pf-c-slider__steps--FontSize: var(--pf-global--FontSize--sm);
+  --pf-c-slider__steps--Height: var(--pf-c-slider__steps--FontSize);
+
+  // step tick
+  --pf-c-slider__step-tick--Top: var(--pf-global--spacer--md);
+  --pf-c-slider__step-tick--Width: #{pf-size-prem(4px)};
+  --pf-c-slider__step-tick--Height: #{pf-size-prem(4px)};
```
Comment (christiemolloy) [inline src/patternfly/components/Slider/slider.scss:18]: should the value for height be the variable for width since they're proportional

File: src/patternfly/components/Slider/slider.scss [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+  // step label
+  --pf-c-slider__step-label--TranslateX: -50%;
+  --pf-c-slider__step-label--Top: calc(var(--pf-global--spacer--xl) + var(--pf-c-slider__rail-track--Height));
+  --pf-c-slider__step--first-child__step-label--TranslateX: 0;
+  --pf-c-slider__step--last-child__step-label--TranslateX: -100%;
+
+  // thumb
+  --pf-c-slider__thumb--Top: calc(var(--pf-c-slider__rail-track--Height) / 2 + var(--pf-global--spacer--md));
+  --pf-c-slider__thumb--Width: #{pf-size-prem(16px)};
```
Comment (christiemolloy) [inline src/patternfly/components/Slider/slider.scss:33]: should the value for height be the variable for width since they're proportional

**Review by christiemolloy — CHANGES_REQUESTED** [review]

File: src/patternfly/components/Slider/examples/Slider.md (hunk 6 above)
Comment (mcoker) [reply src/patternfly/components/Slider/examples/Slider.md:14]: `aria-valuenow` - https://www.w3.org/TR/wai-aria-practices/examples/slider/slider-1.html#rps_label

Though it's just an internal handlebars var so we can call it whatever you'd like to change it.

File: src/patternfly/components/Slider/examples/Slider.md (hunk 7 above)
Comment (mcoker) [reply src/patternfly/components/Slider/examples/Slider.md:15]: It defines:

1. the track fill gradient stops to create the blue fill color - https://github.com/patternfly/patternfly/blob/a296e201f88e918a7ac0e335ec65c95b8b9978d2/src/patternfly/components/Slider/slider.scss#L9
1. positions the thumb - https://github.com/patternfly/patternfly/blob/a296e201f88e918a7ac0e335ec65c95b8b9978d2/src/patternfly/components/Slider/slider.scss#L36
1. positions the floating value input that displays over the thumb - https://github.com/patternfly/patternfly/blob/a296e201f88e918a7ac0e335ec65c95b8b9978d2/src/patternfly/components/Slider/slider.scss#L57

Basically all parts of the UI that update as the value of the slider changes so that the react component only has to update a single var and the slider CSS handles the layout.

File: src/patternfly/components/Slider/examples/Slider.md (hunk 8 above)
Comment (mcoker) [reply src/patternfly/components/Slider/examples/Slider.md:20]: The left position will always have to be a percentage even if the value of the step/tick/label is a different number, because it's used to declare the `left` property in the layout, so it will need to be a percentage.

File: src/patternfly/components/Slider/slider.scss (hunk 9 above)
Comment (mcoker) [reply src/patternfly/components/Slider/slider.scss:18]: It could, though it creates the potential breaking change where, if we make the track 10px tall and we want the tick to be 4x10, if we change `--pf-c-slider__step-tick--Height: var(--pf-c-slider__step-tick--Width);` to `--pf-c-slider__step-tick--Height: 10px;`, then since `--pf-c-slider__step-tick--Width` used to control `--pf-c-slider__step-tick--Height`, and it no longer does, that could be a breaking change. Similar to how we have hover/focus/active, and focus/active aren't defined by the hover var, even though they're all the same value.

Though in this case, I don't think it's likely, but it doesn't cost much to keep them independent.

File: src/patternfly/components/Slider/slider.scss (hunk 10 above)
Comment (mcoker) [reply src/patternfly/components/Slider/slider.scss:34]: Same idea as https://github.com/patternfly/patternfly/pull/3711#discussion_r534333133

**Comment by mcoker** [comment]
@maryshak1996 @christiemolloy updated the thumb color, outline, active/filled tick colors, and made the ticks circles.

**Review by jessiehuff — APPROVED** [review]
This looks good to me! Great job on this!! 😄 🎉

File: src/patternfly/components/Slider/examples/Slider.md (hunk 6 above)
Comment (christiemolloy) [reply src/patternfly/components/Slider/examples/Slider.md:14]: oh right, thats good with me didnt realize it was for aria

File: src/patternfly/components/Slider/slider.scss (hunk 9 above)
Comment (christiemolloy) [reply src/patternfly/components/Slider/slider.scss:18]: ok makes sense lets keep them

File: src/patternfly/components/Slider/slider-main.hbs [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<div class="pf-c-slider__main{{#if slider-main--modifier}} {{slider-main--modifier}}{{/if}}"
```
Comment (christiemolloy) [inline src/patternfly/components/Slider/slider-main.hbs:1]: i think you forgot to add __main to the docs

File: src/patternfly/components/Slider/slider.scss [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+  --pf-c-slider__thumb--Height: #{pf-size-prem(16px)};
+  --pf-c-slider__thumb--Left: var(--pf-c-slider--value);
+  --pf-c-slider__thumb--BackgroundColor: var(--pf-global--primary-color--100);
+  --pf-c-slider__thumb--TranslateX: -50%;
+  --pf-c-slider__thumb--TranslateY: -50%;
+  --pf-c-slider__thumb--BorderRadius: var(--pf-global--BorderRadius--lg);
+  --pf-c-slider__thumb--BoxShadow--base:
+    0 0 0 2px #fff,
```
Comment (christiemolloy) [inline src/patternfly/components/Slider/slider.scss:42]: instead of #fff should this use the variable we have for white

Comment (mcoker) [reply src/patternfly/components/Slider/slider.scss:42]: doh!

**Comment by mcoker** [comment]
@christiemolloy updated

File: src/patternfly/components/Slider/slider.scss (hunk 1 above)
Comment (mattnolting) [reply src/patternfly/components/Slider/slider.scss:54]: Screen Shot 2020-12-02 at 4 01 04 PM

File: src/patternfly/components/Slider/examples/Slider.md (hunk 2 above)
Comment (mattnolting) [reply src/patternfly/components/Slider/examples/Slider.md:178]: Screen Shot 2020-12-02 at 4 01 55 PM

**Review by mattnolting — APPROVED** [review]
🚢  it! 🎉

**Review by christiemolloy — APPROVED** [review]
nice job turning this around so quickly

**Review by mcarrano — APPROVED** [review]
Looks great. Thanks @mcoker !
